### PR TITLE
Parquet reader:  Fix issue when using skip_rows on non-nested columns containing nulls

### PR DESCRIPTION
**Pull request by nvdbaranec**
Fixes  https://github.com/rapidsai/cudf/issues/7113

The basics of the fix are:  when reading non-nested columns, always start reading values at the beginning of the data stream and discarding ones that fall before skip_rows.  Previously we were attempting to jump directly to the first value to decode using a bad assumption.  This was possible with nested types because of the preprocess step we run, but not with flat/simple hierarchies.

Also contains some cleanup and const-ification of things.  In particular, I renamed the variable `out_pos` to `src_pos` in `page_state_s`.  The semantics of the variable are: "the position in the source data to read for the current output position", which made `out_pos` a super obscure name.

### CODE REVIEW

File: cpp/src/io/parquet/page_data.cu [hunk 1]
```diff
@@ -1329,44 +1341,43 @@ static __device__ void gpuUpdateValidityOffsetsAndRowIndices(int32_t target_inpu
...
-      // validity bit for thread t might actually represent output value t-6. the correct position
-      // for thread t's bit is cur_value_count. for cuda 11 we could use __reduce_or_sync(), but
-      // until then we have to do a warp reduce.
-      else {
-        warp_valid_mask = WarpReduceOr32(is_valid << thread_value_count);
-      }
+      uint32_t const warp_valid_mask =
+        // for flat schemas, a simple ballot_sync gives us the correct count and bit positions
+        // because every value in the input matches to a value in the output
+        max_depth == 1
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:1354]: Was the inversion of the if here intended?

Comment (nvdbaranec) [reply cpp/src/io/parquet/page_data.cu:1354]: Sort of.  It should have been this way all along.  The computation ends up the same, it's just that doing it with the warp reduce is more expensive. Whoops.

Comment (devavret) [reply cpp/src/io/parquet/page_data.cu:1354]: I'm having difficulty understanding this. Isn't `max_depth` for flat schema 0?

Comment (nvdbaranec) [reply cpp/src/io/parquet/page_data.cu:1354]: No. max_depth is semantically similar to "size".  A simple int column has a depth of 1.  A list<int> is 2, etc.  Depth doesn't mean "depth past the first level" it just means "height of the hierarchy".

File: cpp/tests/io/parquet_test.cpp [hunk 2]
```diff
@@ -1736,6 +1819,123 @@ TEST_F(ParquetReaderTest, UserBounds)
...
+  }
+}
+
+TEST_F(ParquetReaderTest, UserBoundsWithNullsLarge)
+{
+  constexpr int num_rows = 30 * 1000000;
+  srand(7562);
+  auto frand  = []() { return static_cast<float>(rand()) / static_cast<float>(RAND_MAX); };
+  auto valids = cudf::detail::make_counting_transform_iterator(
+    0, [&](int index) { return frand() >= 0.3 ? 1 : 0; });
```
Comment (vuule) [inline cpp/tests/io/parquet_test.cpp:1865]: How about using https://en.cppreference.com/w/cpp/numeric/random/bernoulli_distribution here? Should even be faster.

**Review by devavret — APPROVED** [review]

File: cpp/tests/io/parquet_test.cpp (hunk 2 above)
Comment (nvdbaranec) [reply cpp/tests/io/parquet_test.cpp]: Done

**Review by shwina — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Comment by nvdbaranec** [comment]
@gpucibot merge

**Pull request merged**